datafusion/core/src/datasource/listing/helpers.rs

```diff
@@ -153,225 +151,239 @@ pub fn split_files(
         .collect()
 }
 
+struct Partition {
+    path: Path,
+    depth: usize,
+    files: Option<Vec<ObjectMeta>>,
+}
+
+impl Partition {
+    async fn list(mut self, store: &dyn ObjectStore) -> Result<(Self, Vec<Path>)> {
+        trace!("Listing partition {}", self.path);
+        let prefix = Some(&self.path).filter(|p| !p.as_ref().is_empty());
+        let result = store.list_with_delimiter(prefix).await?;
+        self.files = Some(result.objects);
+        Ok((self, result.common_prefixes))
+    }
+}
+
+/// Returns a recursive list of the partitions in `table_path` up to `max_depth`
+async fn list_partitions(
+    store: &dyn ObjectStore,
+    table_path: &ListingTableUrl,
+    max_depth: usize,
+) -> Result<Vec<Partition>> {
+    let partition = Partition {
+        path: table_path.prefix().clone(),
+        depth: 0,
+        files: None,
+    };
+
+    let mut out = Vec::with_capacity(64);
+
+    let mut pending = vec![];
+    let mut futures = FuturesUnordered::new();
+    futures.push(partition.list(store));
+
+    while let Some((partition, paths)) = futures.next().await.transpose()? {
+        if let Some(next) = pending.pop() {
```

Comment:
Is there an invariant that if pending is non empty, then futures.len() prior to the loop is `CONCURRENCY_LIMT`? I am trying work out why only one pending future pushed to `futures` rather than pushing while `futures.len() < CONCURRENCY_LIMIT`